File: frontend/src/utils/regions.js
export const ANGLE_MIN = -180;
export const ANGLE_MAX = 180;
export const REGION_COLORS = {
    'alpha-helix': '#4ecdc4',
    'beta-sheet': '#ff6b6b',
    'left-helix': '#45b7d1',
    'disallowed': '#dddddd',
};
export const REGION_LABELS = {
    'alpha-helix': 'α-螺旋',
    'beta-sheet': 'β-折叠',
    'left-helix': '左手螺旋',
    'disallowed': '禁阻区',
};
/** 默认允许区间，与后端 app/regions.py 的 DEFAULT_REGIONS 保持一致 */
export const DEFAULT_REGIONS = [
    { id: 'alpha-helix', label: 'α-螺旋', group: 'alpha-helix', color: REGION_COLORS['alpha-helix'], phiMin: -100, phiMax: -30, psiMin: -80, psiMax: -10 },
    { id: 'beta-sheet', label: 'β-折叠', group: 'beta-sheet', color: REGION_COLORS['beta-sheet'], phiMin: -180, phiMax: -45, psiMin: 60, psiMax: 180 },
    { id: 'beta-sheet-2', label: 'β-折叠（下）', group: 'beta-sheet', color: REGION_COLORS['beta-sheet'], phiMin: -180, phiMax: -45, psiMin: -180, psiMax: -60 },
    { id: 'left-helix', label: '左手螺旋', group: 'left-helix', color: REGION_COLORS['left-helix'], phiMin: 20, phiMax: 100, psiMin: -40, psiMax: 80 },
];
export function cloneDefaults() {
    return DEFAULT_REGIONS.map(r => ({ ...r }));
}
export const BOUND_FIELDS = ['phiMin', 'phiMax', 'psiMin', 'psiMax'];
export const BOUND_FIELD_LABELS = {
    phiMin: 'φ 下限', phiMax: 'φ 上限', psiMin: 'ψ 下限', psiMax: 'ψ 上限',
};
export function toDrafts(defs) {
    return defs.map(d => ({
        id: d.id, label: d.label, group: d.group, color: d.color,
        phiMin: String(d.phiMin), phiMax: String(d.phiMax),
        psiMin: String(d.psiMin), psiMax: String(d.psiMax),
    }));
}
export function errKey(id, field) {
    return `${id}.${field}`;
}
/**
 * 校验草稿并解析为数值区间。
 * 规则：不能为空、须为数字、须在 [-180, 180]、下限不得大于上限（填反）。
 */
export function validateRegionDrafts(drafts) {
    const errors = {};
    const parsed = drafts.map(d => ({
        id: d.id, label: d.label, group: d.group, color: d.color,
        phiMin: NaN, phiMax: NaN, psiMin: NaN, psiMax: NaN,
    }));
    drafts.forEach((d, i) => {
        for (const f of BOUND_FIELDS) {
            const raw = d[f].trim();
            const key = errKey(d.id, f);
            if (raw === '') {
                errors[key] = '不能为空';
                continue;
            }
            const v = Number(raw);
            if (Number.isNaN(v)) {
                errors[key] = '须为有效数字';
                continue;
            }
            if (v < ANGLE_MIN || v > ANGLE_MAX) {
                errors[key] = `须在 ${ANGLE_MIN} ~ ${ANGLE_MAX} 之间`;
                continue;
            }
            parsed[i][f] = v;
        }
        for (const [loF, hiF] of [['phiMin', 'phiMax'], ['psiMin', 'psiMax']]) {
            const lo = parsed[i][loF], hi = parsed[i][hiF];
            if (!Number.isNaN(lo) && !Number.isNaN(hi) && lo > hi) {
                const msg = `下限(${lo})大于上限(${hi})，上下限填反`;
                if (!errors[errKey(d.id, loF)])
                    errors[errKey(d.id, loF)] = msg;
                if (!errors[errKey(d.id, hiF)])
                    errors[errKey(d.id, hiF)] = msg;
            }
        }
    });
    return { errors, parsed };
}

File: backend/app/regions.py
"""Ramachandran 区域判定口径：默认区间、校验与分类的唯一实现。

采样(/api/sample)与重算(/api/reclassify)共用本模块，保证同一份口径。
"""

ANGLE_MIN = -180.0
ANGLE_MAX = 180.0

# 默认允许区间；禁阻区(disallowed)为所有允许区之外的部分
DEFAULT_REGIONS = [
    {"id": "alpha-helix",  "group": "alpha-helix", "phiMin": -100.0, "phiMax": -30.0, "psiMin": -80.0,  "psiMax": -10.0},
    {"id": "beta-sheet",   "group": "beta-sheet",  "phiMin": -180.0, "phiMax": -45.0, "psiMin": 60.0,   "psiMax": 180.0},
    {"id": "beta-sheet-2", "group": "beta-sheet",  "phiMin": -180.0, "phiMax": -45.0, "psiMin": -180.0, "psiMax": -60.0},
    {"id": "left-helix",   "group": "left-helix",  "phiMin": 20.0,   "phiMax": 100.0, "psiMin": -40.0,  "psiMax": 80.0},
]

_FIELDS = (("phi", "phiMin", "phiMax"), ("psi", "psiMin", "psiMax"))


def validate_regions(regions) -> list:
    """校验区间定义，返回错误信息列表（空列表表示合格）。

    规则：上下限不能为空、须落在 [-180, 180]、下限不得大于上限。
    """
    errors = []
    for r in regions:
        rid = r.get("id", "?")
        for axis, lo_key, hi_key in _FIELDS:
            lo, hi = r.get(lo_key), r.get(hi_key)
            if lo is None or hi is None:
                errors.append(f"{rid}: {axis} 上下限不能为空")
                continue
            if not (ANGLE_MIN <= lo <= ANGLE_MAX and ANGLE_MIN <= hi <= ANGLE_MAX):
                errors.append(f"{rid}: {axis} 上下限须位于 [{ANGLE_MIN:g}, {ANGLE_MAX:g}]")
            elif lo > hi:
                errors.append(f"{rid}: {axis} 下限({lo:g})大于上限({hi:g})")
    return errors


def classify_region(phi: float, psi: float, regions=None) -> str:
    """按给定区间判定 (phi, psi) 所属区域；不在任何允许区内则为 disallowed。"""
    for r in (regions or DEFAULT_REGIONS):
        if r["phiMin"] <= phi <= r["phiMax"] and r["psiMin"] <= psi <= r["psiMax"]:
            return r["group"]
    return "disallowed"
